File: scan_hardcoded.dart
import 'dart:io';
import 'dart:convert';

void main() async {
  final baseDir = Directory.current.path;
  final dirsToScan = [
    Directory('$baseDir/lib/features'),
    Directory('$baseDir/lib/core'),
  ];
  final arbArPath = '$baseDir/lib/l10n/app_ar.arb';
  final arbEnPath = '$baseDir/lib/l10n/app_en.arb';

  Map<String, dynamic> loadArb(String path) {
    try {
      final file = File(path);
      if (file.existsSync()) {
        final content = file.readAsStringSync();
        return jsonDecode(content) as Map<String, dynamic>;
      }
    } catch (e) {
      // Ignore
    }
    return {};
  }

  final arbAr = loadArb(arbArPath);
  final arbEn = loadArb(arbEnPath);

  final arValues = <String, String>{};
  final enValues = <String, String>{};

  arbAr.forEach((key, value) {
    if (!key.startsWith('@')) {
      arValues[value.toString().trim()] = key;
    }
  });

  arbEn.forEach((key, value) {
    if (!key.startsWith('@')) {
      enValues[value.toString().trim()] = key;
    }
  });

  final uiParamsPattern = RegExp(
    r"(?:Text\s*\(\s*|title\s*:\s*|label\s*:\s*|hint\s*:\s*|hintText\s*:\s*|labelText\s*:\s*|tooltip\s*:\s*|buttonText\s*:\s*).*?(['"
    '"])(.*?[^\\\\]?)\\1',
  );

  final arabicPattern = RegExp(
    r"(['"
    '"])(.*?[^\\\\]?)\\1',
  );
  final arabicCharPattern = RegExp(r'[\u0600-\u06FF]');

  String generateKey(String text, bool isArabic, int counter) {
    if (!isArabic) {
      final cleanText = text.replaceAll(RegExp(r'[^a-zA-Z0-9 ]'), '');
      final words = cleanText.split(' ').where((w) => w.isNotEmpty).toList();
      if (words.isEmpty) return 'uiKey_$counter';

      String key = words[0].toLowerCase();
      for (int i = 1; i < words.length && i < 5; i++) {
        key += words[i][0].toUpperCase() + words[i].substring(1).toLowerCase();
      }
      return key;
    } else {
      return 'arabicString_$counter';
    }
  }

  final results = <String>[];
  final missingKeysDict = <String, dynamic>{};

  int filesScanned = 0;
  final filesWithStrings = <String>{};
  int totalStringsFound = 0;
  int alreadyInArb = 0;
  int missingFromArb = 0;
  int counter = 1;

  for (final dir in dirsToScan) {
    if (!dir.existsSync()) continue;

    final entities = dir.listSync(recursive: true);
    for (final entity in entities) {
      if (entity is File && entity.path.endsWith('.dart')) {
        filesScanned++;
        final lines = entity.readAsLinesSync();

        bool inBuild = false;
        int braceLevel = 0;

        for (int i = 0; i < lines.length; i++) {
          final line = lines[i];
          final cleanLine = line.trim();

          if (cleanLine.startsWith('//')) continue;

          if (cleanLine.contains('Widget build(')) {
            inBuild = true;
            braceLevel = 0;
          }

          if (inBuild) {
            braceLevel +=
                '{'.allMatches(line).length - '}'.allMatches(line).length;
            if (braceLevel <= 0 && line.contains('}')) {
              inBuild = false;
            }
          }

          final foundStrings = <Map<String, dynamic>>[];

          // Check Arabic strings
          for (final match in arabicPattern.allMatches(line)) {
            final text = match.group(2) ?? '';
            if (arabicCharPattern.hasMatch(text)) {
              foundStrings.add({'text': text, 'isArabic': true});
            }
          }

          // Check English UI strings
          for (final match in uiParamsPattern.allMatches(line)) {
            final text = match.group(2) ?? '';
            if (text.isEmpty || text.trim().length < 2) continue;
            if (arabicCharPattern.hasMatch(text)) continue;
            if (!RegExp(r'[A-Za-z]').hasMatch(text)) continue;
            if (text.startsWith(r'$')) continue;

            foundStrings.add({'text': text, 'isArabic': false});
          }

          final seen = <String>{};
          for (final item in foundStrings) {
            final text = item['text'] as String;
            final isAr = item['isArabic'] as bool;

            if (seen.contains(text)) continue;
            seen.add(text);

            if (text.startsWith('/') ||
                text.startsWith('assets/') ||
                text.endsWith('.png')) {
              continue;
            }

            totalStringsFound++;
            filesWithStrings.add(entity.path);

            bool isInArb = false;
            String suggestedKey = '';

            if (isAr) {
              if (arValues.containsKey(text)) {
                isInArb = true;
                suggestedKey = arValues[text]!;
              }
            } else {
              if (enValues.containsKey(text)) {
                isInArb = true;
                suggestedKey = enValues[text]!;
              }
            }

            if (isInArb) {
              alreadyInArb++;
            } else {
              missingFromArb++;
              suggestedKey = generateKey(text, isAr, counter);
              counter++;

              if (isAr) {
                missingKeysDict[suggestedKey] = "[ARABIC] $text";
                missingKeysDict['@$suggestedKey'] = {
                  "description": "Auto-extracted Arabic string",
                };
              } else {
                missingKeysDict[suggestedKey] = text;
                missingKeysDict['@$suggestedKey'] = {
                  "description": "Auto-extracted English string",
                };
              }
            }

            final relPath = entity.path
                .replaceFirst(baseDir, '')
                .replaceAll('\\', '/');
            results.add(
              "FILE: $relPath\n"
              "LINE: ${i + 1}\n"
              "STRING: \"$text\"\n"
              "SUGGESTED_KEY: $suggestedKey\n"
              "ALREADY_IN_ARB: ${isInArb ? 'YES' : 'NO'}\n",
            );
          }
        }
      }
    }
  }

  final reportFile = File('$baseDir/scan_report.txt');
  final buffer = StringBuffer();
  for (final res in results) {
    buffer.writeln(res);
  }
  buffer.writeln("========================================");
  buffer.writeln("SUMMARY:");
  buffer.writeln("Total files scanned: $filesScanned");
  buffer.writeln("Files with hardcoded strings: ${filesWithStrings.length}");
  buffer.writeln("Total hardcoded strings found: $totalStringsFound");
  buffer.writeln("Already in ARB: $alreadyInArb");
  buffer.writeln("Missing from ARB: $missingFromArb");
  buffer.writeln("========================================");
  buffer.writeln("MISSING KEYS JSON BLOCK:");
  buffer.writeln(const JsonEncoder.withIndent('  ').convert(missingKeysDict));

  reportFile.writeAsStringSync(buffer.toString());
}

File: scan_hardcoded.py
import os
import re
import json

BASE_DIR = r"c:\Users\AL-Reada\Desktop\omar khaled alkhawga\My_App\flutter_project"
DIRS_TO_SCAN = [os.path.join(BASE_DIR, "lib", "features"), os.path.join(BASE_DIR, "lib", "core")]
ARB_AR_PATH = os.path.join(BASE_DIR, "lib", "l10n", "app_ar.arb")
ARB_EN_PATH = os.path.join(BASE_DIR, "lib", "l10n", "app_en.arb")

def load_arb(filepath):
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception:
        return {}

arb_ar = load_arb(ARB_AR_PATH)
arb_en = load_arb(ARB_EN_PATH)

ar_values = {str(v).strip(): k for k, v in arb_ar.items() if not k.startswith('@')}
en_values = {str(v).strip(): k for k, v in arb_en.items() if not k.startswith('@')}

ui_params_pattern = re.compile(
    r"(?:Text\s*\(\s*|title\s*:\s*|label\s*:\s*|hint\s*:\s*|hintText\s*:\s*|labelText\s*:\s*|tooltip\s*:\s*|buttonText\s*:\s*).*?(['\"])(.*?[^\\]?)\1"
)
arabic_pattern = re.compile(r"(['\"])(.*?[^\\]?)\1")

def generate_key(text, is_arabic, counter):
    if not is_arabic:
        words = re.sub(r'[^a-zA-Z0-9 ]', '', text).split()
        if not words:
            return f"uiKey_{counter}"
        key = words[0].lower() + ''.join(word.capitalize() for word in words[1:5])
        return key
    else:
        return f"arabicString_{counter}"

results = []
missing_keys_dict = {}

files_scanned = 0
files_with_strings = set()
total_strings_found = 0
already_in_arb = 0
missing_from_arb = 0
counter = 1

for scan_dir in DIRS_TO_SCAN:
    if not os.path.exists(scan_dir):
        continue
    for root, _, files in os.walk(scan_dir):
        for file in files:
            if not file.endswith('.dart'):
                continue
            
            filepath = os.path.join(root, file)
            files_scanned += 1
            
            with open(filepath, 'r', encoding='utf-8') as f:
                lines = f.readlines()
                
            in_build = False
            brace_level = 0
            
            for line_no, line in enumerate(lines, 1):
                clean_line = line.strip()
                if clean_line.startswith('//'):
                    continue
                
                if 'Widget build(' in clean_line:
                    in_build = True
                    brace_level = 0
                    
                if in_build:
                    brace_level += line.count('{') - line.count('}')
                    if brace_level <= 0 and '}' in line:
                        in_build = False
                        
                found_strings = []
                
                # Check for Arabic strings anywhere in the line
                for match in arabic_pattern.finditer(line):
                    text = match.group(2)
                    if re.search(r'[\u0600-\u06FF]', text):
                        found_strings.append((text, True))
                        
                # Check for English UI strings
                for match in ui_params_pattern.finditer(line):
                    text = match.group(2)
                    if not text or len(text.strip()) < 2:
                        continue
                    if re.search(r'[\u0600-\u06FF]', text):
                        continue
                    if not re.search(r'[A-Za-z]', text):
                        continue
                    if text.startswith('$'):
                        continue
                    found_strings.append((text, False))
                    
                seen = set()
                for text, is_ar in found_strings:
                    if text in seen:
                        continue
                    seen.add(text)
                    
                    # Also ignore some generic programmatic strings like route names or asset paths
                    if text.startswith('/') or text.startswith('assets/') or text.endswith('.png'):
                        continue
                        
                    total_strings_found += 1
                    files_with_strings.add(filepath)
                    
                    is_in_arb = False
                    suggested_key = ""
                    
                    if is_ar:
                        if text in ar_values:
                            is_in_arb = True
                            suggested_key = ar_values[text]
                    else:
                        if text in en_values:
                            is_in_arb = True
                            suggested_key = en_values[text]
                            
                    if is_in_arb:
                        already_in_arb += 1
                    else:
                        missing_from_arb += 1
                        suggested_key = generate_key(text, is_ar, counter)
                        counter += 1
                        
                        if is_ar:
                            missing_keys_dict[suggested_key] = f"[ARABIC] {text}"
                            missing_keys_dict[f"@{suggested_key}"] = {"description": "Auto-extracted Arabic string"}
                        else:
                            missing_keys_dict[suggested_key] = text
                            missing_keys_dict[f"@{suggested_key}"] = {"description": "Auto-extracted English string"}
                    
                    rel_path = os.path.relpath(filepath, BASE_DIR).replace('\\', '/')
                    results.append(f"FILE: {rel_path}\nLINE: {line_no}\nSTRING: \"{text}\"\nSUGGESTED_KEY: {suggested_key}\nALREADY_IN_ARB: {'YES' if is_in_arb else 'NO'}\n")

with open(os.path.join(BASE_DIR, 'scan_report.txt'), 'w', encoding='utf-8') as f:
    for res in results:
        f.write(res + "\n")
    f.write("=" * 40 + "\n")
    f.write("SUMMARY:\n")
    f.write(f"Total files scanned: {files_scanned}\n")
    f.write(f"Files with hardcoded strings: {len(files_with_strings)}\n")
    f.write(f"Total hardcoded strings found: {total_strings_found}\n")
    f.write(f"Already in ARB: {already_in_arb}\n")
    f.write(f"Missing from ARB: {missing_from_arb}\n")
    f.write("=" * 40 + "\n")
    f.write("MISSING KEYS JSON BLOCK:\n")
    f.write(json.dumps(missing_keys_dict, indent=2, ensure_ascii=False) + "\n")
